ykrt/src/compile/jitc_yk/jit_ir/mod.rs

```diff
@@ -1594,9 +1608,59 @@ impl PtrAddInst {
         ptr.unpack()
     }
 
-    pub(crate) fn offset(&self) -> Operand {
-        let unaligned = std::ptr::addr_of!(self.off);
-        unsafe { std::ptr::read_unaligned(unaligned) }.unpack()
+    pub(crate) fn off(&self) -> i32 {
+        self.off
+    }
+}
+
+/// An instruction that adds a (potentially) dynamic offset to a pointer.
+///
+/// The dynamic value is computed from an element size and a number of elements.
+///
+/// # Semantics
+///
+/// Returns a pointer value that is the result of:
+///  - multiplying the constant element size by the dynamic number of elements, and
+///  - adding the result to the specified pointer.
+///
+/// Following LLVM semantics, the operation is permitted to silently wrap if the result doesn't fit
+/// in the LLVM pointer indexing type.
+#[derive(Clone, Debug, PartialEq)]
+#[repr(packed)]
+pub struct DynPtrAddInst {
+    /// The pointer to offset
+    ptr: PackedOperand,
+    /// The element size.
+    ///
+    /// Depending upon the platform, LLVM `getelementptr` may allow larger element sizes than what
+    /// this field can express. Due to space constraints we only accept a reduced range of values
+    /// and traces requiring values outside of this range cannot be JIT compiled.
+    elem_size: u16,
+    /// The (dynamic) number of elements.
+    num_elems: PackedOperand,
+}
+
+impl DynPtrAddInst {
+    pub(crate) fn new(ptr: Operand, num_elems: Operand, elem_size: u16) -> Self {
+        Self {
+            ptr: PackedOperand::new(&ptr),
+            elem_size,
+            num_elems: PackedOperand::new(&num_elems),
+        }
+    }
+
+    pub(crate) fn ptr(&self) -> Operand {
+        let ptr = self.ptr;
```

Comment:
I've just learned that if you want to do an unaligned read of a copyable type, you can do so without using `unsafe` like this.

Note that merging the lines wouldn't compile.